### ISSUE TITLE
Plugin policy not being applied on Windows

### ISSUE THREAD

**Issue by costin** [elastic/elasticsearch#15062]
Due to a particularity in dealing with drive letters, the security policy files for plugins are read but never matched on Windows.

After much hairy pulling, the issue was discovered to be caused in `Security.java#getPluginPermissions()` in particular
`codebases.add(jar.toRealPath().toUri().toURL());`

For the given `Path` which points to `e:\tmp\repository-hdfs\elasticsearch-3.0.0-SNAPSHOT\plugins\lang-groovy\groovy-all-2.4.4-indy.jar`, we get the following behaviour:

`jar.toAbsolutePath()` -> `e:\tmp\repository-hdfs\elasticsearch-3.0.0-SNAPSHOT\plugins\lang-groovy\groovy-all-2.4.4-indy.jar`

`jar.toRealPath()` -> `E:\tmp\repository-hdfs\elasticsearch-3.0.0-SNAPSHOT\plugins\lang-groovy\groovy-all-2.4.4-indy.jar`

Thus the policy is keyed under a upper case `E` but the `ProtectionDomain`/`CodeBase` is using the lower case version:

```
access: domain that failed ProtectionDomain  (file:/e:/tmp/repository-hdfs/elasticsearch-3.0.0-SNAPSHOT/plugins/lang-groovy/groovy-all-2.4.4-indy.jar <no signer certificates>)
 java.net.FactoryURLClassLoader@52af26ee
 <no principals>
 java.security.Permissions@7d30007d (
 ("java.io.FilePermission" "\e:\tmp\repository-hdfs\elasticsearch-3.0.0-SNAPSHOT\plugins\lang-groovy\groovy-all-2.4.4-indy.jar" "read")
)
```

Unfortunately the `URLClassLoader` relies on sun specific classes in retrieving the URL but it looks like it does not rely on the `java.nio` package. Potential solutions would be using `absolutePath` instead though that does not handle symlinks or adding a special check for Windows (which likely might actually work only on Oracle/Sun JDK as it is platform and implementation specific) that lowercases the drive letter.
I wonder who this behaves with things like SMB or other crazy URLs.

**Comment by rmuir** [comment]
Thanks Costin. Unfortunately we need symlink resolution... We got eg jenkins boxes with linked .m2. And despite the trouble i think simple string lookup is important. My current thinking is .toFile then getCanonical... But i have not looked yet.

**Comment by rmuir** [comment]
Something like this is my idea (needs testing):
```
--- a/core/src/main/java/org/elasticsearch/bootstrap/Security.java
+++ b/core/src/main/java/org/elasticsearch/bootstrap/Security.java
@@ -141,7 +141,9 @@ final class Security {
                         List<URL> codebases = new ArrayList<>();
                         try (DirectoryStream<Path> jarStream = Files.newDirectoryStream(plugin, "*.jar")) {
                             for (Path jar : jarStream) {
-                                codebases.add(jar.toRealPath().toUri().toURL());
+                                // codebase normalization in java is crazy, use java.io.File on purpose,
+                                // since that is what is being used by classloaders.
+                                codebases.add(jar.toFile().getCanonicalFile().toURI().toURL());
                             }
                         }
```

**Comment by costin** [comment]
@rmuir if it's okay, I'll try to fix this one myself.

**Comment by rmuir** [comment]
Thanks. My only concern is nobody can repro this right now. Jenkins is happy on windows systems it tests... So it would be great to have some explanation why that is.

**Comment by costin** [comment]
Got to the root of the problem and figured out why the tests don't fail on Jenkins.
In my case ES and Eclipse are started many times from a file explorer (TotalCommander in this case). Which in turn changes the drive letter in the spawn consoles (like `cmd`) to be lowercase.
Windows on the other hand likes upper case drives.
ES starts, picks up the environment and thus thinks it's under `e:\...` drive (in `path.home`). Everything else that is relative to this will use the `e:\` in its path. However when calling `toRealPath`, the file-system and not the environment is checked and the upper case drive letter is used instead (`E:\`).

After stopping the explorer to change the drive letter and use the upper case one instead, things worked out fine out of the box.
So in short, this is not a bug but rather an error in the user environment. It looks to be Windows specific, maybe there's a way to do some sanity checks in the `.bat` file? /cc @gmarz @Mpdreamz

**Comment by costin** [comment]
Closing it down.
